awx/main/dispatch/periodic.py

```diff
@@ -1,57 +1,72 @@
-                if os.getppid() != ppid:
-                    # if the parent PID changes, this process has been orphaned
-                    # via e.g., segfault or sigkill, we should exit too
-                    pid = os.getpid()
-                    logger.warning(f'periodic beat exiting gracefully pid:{pid}')
-                    raise SystemExit()
-                try:
-                    for conn in connections.all():
-                        # If the database connection has a hiccup, re-establish a new
-                        # connection
-                        conn.close_if_unusable_or_obsolete()
-                    set_guid(generate_guid())
-                    self.run_pending()
-                except Exception:
-                    logger.exception('encountered an error while scheduling periodic tasks')
-                time.sleep(idle_seconds)
-
-        process = Process(target=run)
-        process.daemon = True
-        process.start()
-
-
-def run_continuously():
-    scheduler = Scheduler()
-    for task in settings.CELERYBEAT_SCHEDULE.values():
-        apply_async = TaskWorker.resolve_callable(task['task']).apply_async
-        total_seconds = task['schedule'].total_seconds()
-        scheduler.every(total_seconds).seconds.do(apply_async)
-    scheduler.run_continuously()
+class Job:
+    def __init__(self, name: str, data: dict):
+        self.name = name
+        self.data = data
+        self.interval = int(data['schedule'].total_seconds())
+        self.offset = None
+        self.index = 0
+        self.missed_runs = 0
+
+    @property
+    def next_run(self):
+        """
+        Gives the time until the next run with t=0 being the global_start
+        of the scheduler class
+        """
+        return (self.index + 1) * self.interval + self.offset
+
+    def mark_run(self, relative_time):
+        new_index = (relative_time - self.offset) // self.interval
+        if new_index > self.index + 1:
+            logger.warning(f'Missed {new_index - self.index} schedules of {self.name}')
+            self.missed_runs += 1
+        self.index = new_index
+
+
+class Scheduler:
+    def __init__(self, schedule):
+        """
+        Expects a schedule in the form of a dictionary like
+        {'job1': {'task': 'foo.bar', 'interval': timedelta(seconds=50)}}
+        The goal is to return pending tasks at a given time and time until
+        the next job to be run.
+        The keys are ignored, only the inverval from the values are used.
+        """
+        self.jobs = [Job(name, data) for name, data in schedule.items()]
+        min_interval = min(job.interval for job in self.jobs)
+        num_jobs = len(self.jobs)
+
+        # even space out jobs over the base interval
+        for i, job in enumerate(self.jobs):
+            job.offset = (i * min_interval) // num_jobs
+
+        # internally times are all referenced relative to startup time, add grace period
+        self.global_start = int(time.time() + 2)
```

Comment:
I've thought about this a lot, and changed my mind that `global_start` should stay as a float. Our "internal" time model should remain all as integer math, but we shouldn't pin the target run times to on-the-second times, because if all applications did this, it would create artificial bottlenecks.